src/Castle.Facilities.AspNet.Mvc/AspNetMvcControllerFactory.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2004-2017 Castle Project - http://www.castleproject.org/
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Reflection;
+
+namespace Castle.Facilities.AspNet.Mvc
+{
+	using System;
+	using System.Collections.Generic;
+	using System.Linq;
+	using System.Web;
+	using System.Web.Mvc;
+	using System.Web.Routing;
+	using System.Web.SessionState;
+
+	using Castle.Facilities.AspNet.Mvc.Exceptions;
+	using Castle.MicroKernel;
+	using Castle.MicroKernel.Lifestyle;
+
+	public class AspNetMvcControllerFactory : IControllerFactory
+	{
+		private const string ContextScopeKey = "castle-windsor-facility-aspnet-mvc-scope";
+
+		private readonly ISupportFacility facilitySupport;
+		private List<Assembly> controllerAssemblies;
+
+		public event Action<IKernel, Type> BeforeControllerResolved;
+		public event Action<IKernel, object> AfterControllerReleased;
+
+		public AspNetMvcControllerFactory(ISupportFacility facilitySupport)
+		{
+			this.facilitySupport = facilitySupport;
+		}
+
+		public IController CreateController(RequestContext requestContext, string controllerName)
+		{
+			var adjustedControllerName = $"{controllerName}Controller";
+			var controllerType = FindControllerType(adjustedControllerName);
+			BeginScopedLifestyleIfRequired();
+			OnBeforeControllerResolved(controllerType);
+			try
+			{
+				return (IController) facilitySupport.Kernel.Resolve(controllerType);
+			}
+			catch (InvalidOperationException err)
+			{
+				throw new LifestyleScopesPotentiallyNotEnabledException("If you are having trouble with scoped lifestyles, please try 'container.AddFacility<AspNetMvcFacility>(x => x.AddControllerAssembly<MvcApplication>().WithLifestyleScopedPerWebRequest())'", err);
+			}
+		}
+
+		public SessionStateBehavior GetControllerSessionBehavior(RequestContext requestContext, string controllerName)
+		{
+			return SessionStateBehavior.Default;
```

Comment:
I don't know what this stuff does but do we not care if someone applies an `SessionStateAttribute` to their controller?

Reply:
Yes, you are totally right we should follow best practice and set this to disabled.

Reply:
Just kidding, we might need to pass this as a parameter via the public api. Looks like the IControllerFactory demands it for now.

Reply:
Additional stuff about what it gives you. You basically end up with a Session State object. It tracks the browser and reconciles http/s requests with a user `session` which then allows you to store user profile information that is retrievable on each subsequent request using a dictionary basically. I have seen this abused also, projects where people try and cache the entire database in this thing where worker processes very quickly jump to 2GB.  You can also easily back this store onto something like redis or sql very easily luckily.

Reply:
Addressed in commit: 9c8aa17

Reply:
Don't worry I know what session and view state are, and also seem them heavily abused in the wild. I just had no idea what this enum does here, it sort of sounds like the "handler" (i.e. the controller factory) is the one telling ASP.NET what to do as the fallback, and this is per controller.